pkg/controller/build/build_controller.go

```diff
@@ -99,6 +100,96 @@ func add(ctx context.Context, mgr manager.Manager, r reconcile.Reconciler) error
 
 	// Watch for changes to primary resource Build
 	err = c.Watch(&source.Kind{Type: &build.Build{}}, &handler.EnqueueRequestForObject{}, pred)
+
+	preSecret := predicate.Funcs{
+
+		// Only filter events where the secret have the Build specific annotation
+		CreateFunc: func(e event.CreateEvent) bool {
+			objectAnnotations := e.Meta.GetAnnotations()
+			if _, ok := buildSecretRefAnnotationExist(objectAnnotations); ok {
+				return true
+			}
+			return false
```

Comment:
Our past observation was that a `CreateEvent` happens for all existing entities in the system when the controller is taking over leadership. At that point, all builds are reconciled anyway. It does not make sense to look at secrets in such a scenario as this would only cause additional load on the system. Should we therefore add a check here to filter out objects that are older than, for example, one minute ?

Reply:
@SaschaSchwarze0 I think below part can eliminate your concern:
```
if len(secretNames) > 0 {
		if err := r.validateSecrets(ctx, secretNames, b.Namespace); err != nil {
			b.Status.Reason = err.Error()
			if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
				return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
			}
			// The Secret Resource watcher will Reconcile again once the missing
			// secret is created, therefore no need to return an error and enter on an infinite
			// reconciliation
			return reconcile.Result{}, nil
		}
	}
```
If my understanding is correct, then yes, when controller leader switches, all entities will trigger a create event. And all builds will reconcile. But at that time, for these builds, they are referencing secrets don't exist, their reconciles are triggered via a secret create event rather than leader switches. So I don't think this will cause additional load.  

And current logic is if a build failed because secret doesn't exist, then this build will stop reconciling until a secret create event happens.

@qu1queee If I am wrong, pls correct me, thanks!

Reply:
@SaschaSchwarze0 thanks, I actually needed to test this to verify it, which is the case. I think this types of improvements we should do them on a bigger card, something via changing our HA model from active/passive to active/active. Please read the internal doc we have around the tekton active/active, they explain there an approach to warm up the pods in advance(informers cache), which should address this concern.

Reply:
I like active/active. In that concept I was missing information on how the situation will be handled that all instances of the controller are down and then need to catch up on changes that happened while they were not watched. But that detail might have been somewhere and I oversaw it.

I would also like if we switch to that model at some point.

But let me try to describe the situation how I understand it with an example. Scenario is:

- one namespace with ten builds and three secrets (A, B and C)
- for simplicity all secrets are related to builds and are annotated
- five builds reference a container registry secret A for the output image
- the other five reference a container registry secret B for the output image
- five builds also reference the same git source secret C

When our controller takes over leadership, the existing code will do this:

- reconcile ten builds

The new code will do this:

- reconcile ten builds
- check three secrets, for each secret (because it is annotated) it will:
  - list the builds in the namespace and filter them, in every case, five builds are found
  - reconcile the five builds

So, the overhead is three build list calls and 15 build reconciliations.

-- 

I think that the build custom resource is the critical resource when it comes to controller starts, especially in Code Engine:

1. There is a fixed and low number of cluster build strategies that are reconciled
2. Build strategies do not exist at all in Code Engine, I have not thought about how frequently they will be used elsewhere
3. Build runs is likely the entity which exists the most compared to everything else from us, but for build runs we have a predicate function that filters our completed build runs from reconciliation

So, if my calculation from above is accurate, I have a little stomach ache that - with the current implementation - we are making things really worse on leadership change. I would feel better if it would look like this:

```go
		CreateFunc: func(e event.CreateEvent) bool {
			// Only consider CreateEvents that just happened. A CreateEvent with a CreationTimestamp in the past happens
			// when the controller is started. In this case, all builds are reconciled anyway and there is no reason in
			// reconciling them through the referenced secret again
			if e.Meta.GetCreationTimestamp().Add(1 * time.Minute).After(time.Now()) {
				// Only filter events where the secret have the Build specific annotation
				objectAnnotations := e.Meta.GetAnnotations()
				if _, ok := buildSecretRefAnnotationExist(objectAnnotations); ok {
					return true
				}
			}
			return false
		},
```

Reply:
It took me some time to find out why this filtering(_avoiding reconcile secrets that are older than 1 minute_) approach smelled fishy for me, let me share my findings via two facts:

_Note:_ I think you are right on the fact that this will allow us to avoid doing some API calls.

**Fact 1.** The use-case layout here have a technical explanation(I learned the technicalities today) although I saw the behaviour before. You can read more about the official behaviour from https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes, and I quote it:

> If the client watch is disconnected they can restart a new watch from the last returned resourceVersion, or perform a new collection request and begin again. See Resource Version Semantics for more detail.

From the above, the `watch is disconnected` is basically when our current operator is down and we restarted or get a new pod running the same binary. Now, the part where they mentioned `can restart a new watch from the last returned resourceVersion` is pretty important, because if a controller is watching a collection of resources (_in our case secrets_), whenever the controller is restarted, the controller should be able to get access to the latest resourceVersion of each secret. 

You can read more on the resourceVersion in here https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions, but the problem I see with `avoiding reconcile secrets that are older than 1 minute`, is that during the time the controller was down, and old secret might have gotten an update and its resourceVersion have changed. Once the controller restarts, it would not be able to watch the last return resourceVersion, which feels wrong because we want controllers in general to get the latest state of a resource(as seen from the DB) and try to match this with the current state of that object in the cluster(core k8s concept). Now, this of course in the context of the Build/Secret/Reconciles have none implications, but in general, allowing this type of operations in our code could lead to very bad practices in the future for other types of operations.

As an exercise, while you controller is stopped, you can always modify a secret and verify the changes in the resourceVersion, for example:
```sh
$ kubectl -n test-build get secret icr-knbuild -o json | jq -r .metadata.resourceVersion
43361

$ kubectl -n test-build annotate secrets icr-knbuild build.build.dev/foobarchange='true'
secret/icr-knbuild annotated

$ kubectl -n test-build get secret icr-knbuild -o json | jq -r .metadata.resourceVersion
49363
```

Now, once you restart your operator, you could print the secret resourceVersion in the CREATE predicate(see [example](https://github.com/shipwright-io/build/commit/cef3781d4bdcff2d951394c5ec3e9fb577c39245)), and you will see that the controller gets a `49363` after restarting. Again, not allowing the controller to observe the latest version of a resource during restart feels wrong.


**Fact 2.** Basically Idempotency. 

In a nutshell we want our controller to be idempotent, which basically means that no matter the amount of times we reconcile, there should not be additional updates on our objects if nothing changed. This seems pretty important for me, because instead of concentrating on the amount of times we might reconcile, we should pay more attention on if we are idempotent or not(which I believe is the case). E.g. after a controller is restarted, no matter the amount of reconciles we do on a Build(due to secrets), the Build is not changed, meaning its `metadata.resourceVersion` remains intact. Some good articles on this:
- https://blog.container-solutions.com/a-deep-dive-into-the-java-operator-sdk
- https://stackoverflow.com/questions/1077412/what-is-an-idempotent-operation
- https://github.com/kubernetes-sigs/kubebuilder/issues/618#issuecomment-499705932
- https://cluster-api.sigs.k8s.io/developer/providers/implementers-guide/controllers_and_reconciliation.html#idempotency (this one is concerned around reconciling and doing updates on each reconcile)


So, finding common ground:
- Yes, we can save some API calls by _avoiding reconcile secrets that are older than 1 minute_, but we are hiding information from our controller during the ramp-up after restart. While this might not be a problem now, it can be a problem later on other types of resources, where we might follow this same approach.
- Multiple reconciles are an overhead but we understand they will not end-up with any UPDATEs on our Builds, as long as secrets and Builds integrity didnt change during the downtime of the controller.
- We do have an idempotent Build controller, does this justify the extra API calls during ramp-up time ?

Reply:
Modified secrets where the annotation is added is indeed the scenario I was searching for but not finding that requires to reconcile all secrets with the annotation on `CreateEvent`.